File: 633SumofSquareNumbers/withTowPointer.c
#include <stdio.h>
#include <math.h>


bool judgeSquareSum(int c){
    long a = 0;
    long b = (long)sqrt(c);

    while(a <= b) {
        long sum = a*a + b*b;
        if(sum == c) {
            return true;
        }

        if(sum < c) {
            a++;
        } else {
            b--;
        }
    }

    return false;
}

int main() {
	printf("%d\n", judgeSquareSum(3));
	printf("%d\n", judgeSquareSum(5));
	printf("%d\n", judgeSquareSum(8));
	printf("%d\n", judgeSquareSum(7));
}

File: 633SumofSquareNumbers/withTowPointer.go
// https://leetcode-cn.com/problems/sum-of-square-numbers/description/

// 双指针解法 (逼近、缩减范围)
// 时间 O(sqrt(c)) (从 0 ~ sqrt(c) 遍历) 空间 O(1)

package main

import (
	"fmt"
	"math"
)

func main() {
	inputs := map[int]bool{
		3: false,
		5: true,
		6: false,
		7: false,
		8: true,
		9: true,
	}

	for input, res := range inputs {
		if judgeSquareSum(input) != res {
			panic("error")
		}
	}

	fmt.Println("Done")
}

func judgeSquareSum(c int) bool {

	if c < 0 {
		return false
	}

	a := 0
	b := int(math.Sqrt(float64(c)))

	for a <= b {
		pow := a*a + b*b
		if pow == c {
			return true
		}

		if pow < c {
			a++
		} else {
			b--
		}
	}

	return false
}
